File: src/apple/Packages/MapleCore/Sources/MapleCore/LocalAdjustmentFlat.swift
// LocalAdjustmentFlat.swift — byte-for-byte Swift mirror of
// raw_core::types::local_adjustment::flat (#3274). 32 Float32 per layer;
// the slot map is documented on that Rust file's header and MUST be kept in
// lockstep with it — a divergence here is a live-vs-fallback rendering bug,
// not a compile error.

import Foundation

public enum LocalAdjustmentFlat {
    public static let layerFloatLen = 32

    private static let kindLinear: Float = 0
    private static let kindRadial: Float = 1
    private static let kindBitmap: Float = 2
    private static let kindEverywhere: Float = 3
    private static let rangeKindColor: Float = 1

    private static let presentExposure: Int = 1 << 0
    private static let presentContrast: Int = 1 << 1
    private static let presentHighlights: Int = 1 << 2
    private static let presentShadows: Int = 1 << 3
    private static let presentWhites: Int = 1 << 4
    private static let presentBlacks: Int = 1 << 5
    private static let presentSaturation: Int = 1 << 6
    private static let presentVibrance: Int = 1 << 7
    private static let presentTemperature: Int = 1 << 8
    private static let presentTint: Int = 1 << 9
    private static let presentHue: Int = 1 << 10

    public static func toFlat(_ layers: [LocalAdjustment]) -> [Float] {
        var out = [Float](repeating: 0, count: layers.count * layerFloatLen)
        for (i, layer) in layers.enumerated() {
            let base = i * layerFloatLen
            writeMask(layer.mask, into: &out, base: base)
            writeAdjustments(layer.adjustments, into: &out, base: base)
            writeRange(layer.range, into: &out, base: base)
        }
        return out
    }

    private static func writeMask(_ mask: LocalMask, into out: inout [Float], base: Int) {
        switch mask {
        case .linear(let start, let end, let feather):
            out[base + 0] = Float(start.x)
            out[base + 1] = Float(start.y)
            out[base + 2] = Float(end.x)
            out[base + 3] = Float(end.y)
            out[base + 4] = Float(feather)
            out[base + 6] = kindLinear
        case .radial(let center, let radii, let angle, let feather, let invert):
            out[base + 0] = Float(center.x)
            out[base + 1] = Float(center.y)
            out[base + 2] = Float(radii.x)
            out[base + 3] = Float(radii.y)
            out[base + 4] = Float(feather)
            out[base + 5] = Float(angle)
            out[base + 6] = kindRadial
            out[base + 7] = invert ? 1 : 0
        case .bitmap(_, let rasterId):
            out[base + 2] = Float(rasterId)
            out[base + 6] = kindBitmap
        case .everywhere:
            out[base + 6] = kindEverywhere
        }
    }

    private static func writeAdjustments(_ a: PartialAdjustments, into out: inout [Float], base: Int) {
        let fields: [(Double?, Int)] = [
            (a.exposure, presentExposure), (a.contrast, presentContrast), (a.highlights, presentHighlights),
            (a.shadows, presentShadows), (a.whites, presentWhites), (a.blacks, presentBlacks),
            (a.saturation, presentSaturation), (a.vibrance, presentVibrance),
            (a.temperature, presentTemperature), (a.tint, presentTint),
        ]
        var present = 0
        for (value, bit) in fields where value != nil { present |= bit }
        if a.hue != nil { present |= presentHue }
        out[base + 8] = Float(present)
        for (i, (value, _)) in fields.enumerated() { out[base + 12 + i] = Float(value ?? 0) }
        out[base + 22] = Float(a.hue ?? 0)
    }

    private static func writeRange(_ range: RangeRefinement?, into out: inout [Float], base: Int) {
        guard case let .color(hueDeg, halfWidth, chromaMin, lMin, lMax, feather) = range else { return }
        out[base + 24] = rangeKindColor
        out[base + 25] = Float(hueDeg)
        out[base + 26] = Float(halfWidth)
        out[base + 27] = Float(chromaMin)
        out[base + 28] = Float(lMin)
        out[base + 29] = Float(lMax)
        out[base + 30] = Float(feather)
    }

    /// `rasterDigests` maps a resolved raster id (slot 2 of a bitmap record)
    /// to its digest, so a decoded `LocalMask.bitmap`'s recipe carries the right
    /// identity even though the flat wire itself only stores the id.
    public static func fromFlat(_ flat: [Float], rasterDigests: [UInt32: String]) -> [LocalAdjustment] {
        stride(from: 0, to: flat.count - flat.count % layerFloatLen, by: layerFloatLen).map { base in
            LocalAdjustment(
                mask: readMask(flat, base: base, rasterDigests: rasterDigests),
                range: readRange(flat, base: base),
                adjustments: readAdjustments(flat, base: base)
            )
        }
    }

    private static func readMask(_ flat: [Float], base: Int, rasterDigests: [UInt32: String]) -> LocalMask {
        let kind = flat[base + 6]
        if kind == kindEverywhere { return .everywhere }
        if kind == kindBitmap {
            let rasterId = UInt32(flat[base + 2])
            let digest = rasterDigests[rasterId] ?? ""
            return .bitmap(
                recipe: BitmapRecipe(person: 0, facialSkin: true, bodySkin: true, model: "", digest: digest),
                rasterId: rasterId
            )
        }
        if kind == kindRadial {
            return .radial(
                center: MaskPoint(x: Double(flat[base + 0]), y: Double(flat[base + 1])),
                radii: MaskPoint(x: Double(flat[base + 2]), y: Double(flat[base + 3])),
                angle: Double(flat[base + 5]), feather: Double(flat[base + 4]), invert: flat[base + 7] != 0
            )
        }
        return .linear(
            start: MaskPoint(x: Double(flat[base + 0]), y: Double(flat[base + 1])),
            end: MaskPoint(x: Double(flat[base + 2]), y: Double(flat[base + 3])), feather: Double(flat[base + 4])
        )
    }

    private static func readAdjustments(_ flat: [Float], base: Int) -> PartialAdjustments {
        let present = Int(flat[base + 8])
        func field(_ i: Int, _ bit: Int) -> Double? { present & bit != 0 ? Double(flat[base + 12 + i]) : nil }
        return PartialAdjustments(
            exposure: field(0, presentExposure), contrast: field(1, presentContrast),
            highlights: field(2, presentHighlights),
            shadows: field(3, presentShadows), whites: field(4, presentWhites), blacks: field(5, presentBlacks),
            saturation: field(6, presentSaturation), vibrance: field(7, presentVibrance),
            temperature: field(8, presentTemperature), tint: field(9, presentTint),
            hue: present & presentHue != 0 ? Double(flat[base + 22]) : nil
        )
    }

    private static func readRange(_ flat: [Float], base: Int) -> RangeRefinement? {
        guard flat[base + 24] == rangeKindColor else { return nil }
        return .color(
            hueDeg: Double(flat[base + 25]), hueHalfWidthDeg: Double(flat[base + 26]),
            chromaMin: Double(flat[base + 27]),
            lMin: Double(flat[base + 28]), lMax: Double(flat[base + 29]), feather: Double(flat[base + 30])
        )
    }
}

File: src/windows/Maple.WinUI/Native/LocalAdjustmentFlat.cs
// LocalAdjustmentFlat — C# mirror of raw-core's flat `f32` wire for a
// local-adjustment layer stack (`raw_core::types::local_adjustment::flat`,
// #1698). One layout serves the FFI `(ptr, len)` pair on both
// MapleAdjustmentParams (CPU chain) and MapleGpuLiveParams (GPU live chain) —
// `local_adjustments_ptr`/`local_adjustments_len`, already present on both
// C# struct mirrors (#3406) but never populated before this file, which is
// why a sidecar authored on web/Apple rendered correctly on Windows (the
// Rust stage runs) while a layer added ON Windows never reached the
// renderer.
//
// This is encode-only: Windows never reads the flat wire back (the sidecar
// round-trip owns reads — Services/Xmp/XmpLocalAdjustments.cs), it only
// serializes the in-memory model for the per-tick FFI call, same as Swift's
// `LocalAdjustmentFlat.swift` and TS's flat encoder.

using System;
using System.Collections.Generic;
using Maple.WinUI.Models;

namespace Maple.WinUI.Native
{
    internal static class LocalAdjustmentFlat
    {
        /// <summary>Floats per serialized layer — eight WGSL vec4&lt;f32&gt;
        /// members, matching raw-core's `LAYER_FLAT_LEN`.</summary>
        public const int LayerFlatLen = 32;

        private const float KindLinear = 0f;
        private const float KindRadial = 1f;

        private const uint PresentExposure = 1u << 0;
        private const uint PresentContrast = 1u << 1;
        private const uint PresentHighlights = 1u << 2;
        private const uint PresentShadows = 1u << 3;
        private const uint PresentWhites = 1u << 4;
        private const uint PresentBlacks = 1u << 5;
        private const uint PresentSaturation = 1u << 6;
        private const uint PresentVibrance = 1u << 7;
        private const uint PresentTemperature = 1u << 8;
        private const uint PresentTint = 1u << 9;
        private const uint PresentHue = 1u << 10;

        /// <summary>Serialize a layer stack to the flat wire — length is
        /// always <c>layers.Count * LayerFlatLen</c>; an empty stack yields
        /// an empty array (both FFI structs read that as "no local
        /// adjustments," the bit-identical short-circuit raw-core's own
        /// `apply` takes on an empty slice).</summary>
        public static float[] ToFlat(IReadOnlyList<LocalAdjustment> layers)
        {
            var flat = new float[layers.Count * LayerFlatLen];
            for (var i = 0; i < layers.Count; i++)
                WriteLayer(layers[i], flat.AsSpan(i * LayerFlatLen, LayerFlatLen));
            return flat;
        }

        private static void WriteLayer(LocalAdjustment layer, Span<float> slot)
        {
            WriteMask(layer.Mask, slot);
            WriteAdjustments(layer.Adjustments, slot);
            WriteRange(layer.Range, slot);
        }

        private static void WriteMask(LocalMask mask, Span<float> slot)
        {
            switch (mask)
            {
                case LinearMask l:
                    slot[0] = (float)l.Start.X;
                    slot[1] = (float)l.Start.Y;
                    slot[2] = (float)l.End.X;
                    slot[3] = (float)l.End.Y;
                    slot[4] = (float)l.Feather;
                    slot[6] = KindLinear;
                    break;
                case RadialMask r:
                    slot[0] = (float)r.Center.X;
                    slot[1] = (float)r.Center.Y;
                    slot[2] = (float)r.Radii.X;
                    slot[3] = (float)r.Radii.Y;
                    slot[4] = (float)r.Feather;
                    slot[5] = (float)r.Angle;
                    slot[6] = KindRadial;
                    slot[7] = r.Invert ? 1f : 0f;
                    break;
                default:
                    throw new InvalidOperationException($"unknown mask shape {mask.GetType().Name}");
            }
        }

        // A local function closing over `slot` (a ref struct) is not legal
        // C# (CS9108), so this is unrolled rather than looped over a
        // (field, bit) descriptor table the way the Rust encoder does.
        private static void WriteAdjustments(PartialAdjustments a, Span<float> slot)
        {
            var present = 0u;
            if (a.Exposure is { } exposure) { slot[12] = (float)exposure; present |= PresentExposure; }
            if (a.Contrast is { } contrast) { slot[13] = (float)contrast; present |= PresentContrast; }
            if (a.Highlights is { } highlights) { slot[14] = (float)highlights; present |= PresentHighlights; }
            if (a.Shadows is { } shadows) { slot[15] = (float)shadows; present |= PresentShadows; }
            if (a.Whites is { } whites) { slot[16] = (float)whites; present |= PresentWhites; }
            if (a.Blacks is { } blacks) { slot[17] = (float)blacks; present |= PresentBlacks; }
            if (a.Saturation is { } saturation) { slot[18] = (float)saturation; present |= PresentSaturation; }
            if (a.Vibrance is { } vibrance) { slot[19] = (float)vibrance; present |= PresentVibrance; }
            if (a.Temperature is { } temperature) { slot[20] = (float)temperature; present |= PresentTemperature; }
            if (a.Tint is { } tint) { slot[21] = (float)tint; present |= PresentTint; }
            if (a.Hue is { } hue) { slot[22] = (float)hue; present |= PresentHue; }
            slot[8] = present;
        }

        private static void WriteRange(ColorRangeRefinement? range, Span<float> slot)
        {
            if (range is not { } r)
                return; // slot[24] stays 0 == RANGE_KIND_NONE, the array default.
            slot[24] = 1f; // RANGE_KIND_COLOR
            slot[25] = (float)r.HueDeg;
            slot[26] = (float)r.HueHalfWidthDeg;
            slot[27] = (float)r.ChromaMin;
            slot[28] = (float)r.LMin;
            slot[29] = (float)r.LMax;
            slot[30] = (float)r.Feather;
        }
    }
}
